src/ir/possible-contents.cpp

```diff
@@ -102,36 +102,180 @@ void PossibleContents::combine(const PossibleContents& other) {
-    // If only one is a null, but the other's type is known exactly, then the
-    // combination is to add nullability (if the type is *not* known exactly,
-    // like for a global, then we cannot do anything useful here).
-    if (!isNull() && hasExactType()) {
-      value = ExactType(Type(type.getHeapType(), Nullable));
+    // If only one is a null then we can use the type info from the other, and
+    // just add in nullability. For example, a literal of type T and a null
+    // becomes an exact type of T that allows nulls, and so forth.
+    auto mixInNull = [](ConeType cone) {
+      cone.type = Type(cone.type.getHeapType(), Nullable);
+      return cone;
+    };
+    if (!isNull()) {
+      value = mixInNull(getCone());
       return;
-    } else if (!other.isNull() && other.hasExactType()) {
-      value = ExactType(Type(otherType.getHeapType(), Nullable));
+    } else if (!other.isNull()) {
+      value = mixInNull(other.getCone());
       return;
     }
   }
 
-  if (hasExactType() && other.hasExactType() &&
-      type.getHeapType() == otherType.getHeapType()) {
-    // We know the types here exactly, and even the heap types match, but
-    // there is some other difference that prevents them from being 100%
-    // identical (for example, one might be an ExactType and the other a
-    // Literal; or both might be ExactTypes and only one might be nullable).
-    // In these cases we can emit a proper ExactType here, adding nullability
-    // if we need to.
-    value = ExactType(Type(
-      type.getHeapType(),
-      type.isNullable() || otherType.isNullable() ? Nullable : NonNullable));
+  // Before we give up and return Many, try to find a ConeType that describes
+  // both inputs.
+  auto lub = Type::getLeastUpperBound(type, otherType);
+  if (lub != Type::none) {
+    // We found a shared ancestor. Next we need to find how big a cone we need:
+    // the cone must be big enough to contain both the inputs.
+    auto depth = getCone().depth;
+    auto otherDepth = other.getCone().depth;
+    Index newDepth;
+    if (depth == FullDepth || otherDepth == FullDepth) {
+      // At least one has full (infinite) depth, so we know the new depth must
+      // be the same.
+      newDepth = FullDepth;
+    } else {
+      // The depth we need under the lub is how far from the lub we are, plus
+      // the depth of our cone.
+      // TODO: we could make a single loop that also does the LUB, at the same
+      // time, and also avoids calling getDepth() which loops once more?
+      auto depthFromRoot = type.getHeapType().getDepth();
+      auto otherDepthFromRoot = otherType.getHeapType().getDepth();
+      auto lubDepthFromRoot = lub.getHeapType().getDepth();
+      assert(lubDepthFromRoot <= depthFromRoot);
+      assert(lubDepthFromRoot <= otherDepthFromRoot);
+      Index depthUnderLub = depthFromRoot - lubDepthFromRoot + depth;
+      Index otherDepthUnderLub =
+        otherDepthFromRoot - lubDepthFromRoot + otherDepth;
```

Comment:
`getSuperType` only returns user types atm. So it never returns a basic type like `data`. (This might be worth improving, but I looked into it and it wasn't trivial.)

The depth should be unique, and is defined to include basic supertypes in the count.